### DISCUSSION TITLE
Calling your library/class object from Node Red, I ultimately get a timeout?

### DISCUSSION THREAD

**Discussion by Jibun-no-Kage** [WesSouza/lgtv-ip-control#111] [Q&A]
Calling your library/class object from Node Red, I ultimately get a timeout?

TimeoutError
    at Socket.handleTimeout (/root/node_modules/lgtv-ip-control/dist/classes/TinySocket.cjs:38:16)
    at Object.onceWrapper (node:events:631:28)
    at Socket.emit (node:events:517:28)
    at Socket._onTimeout (node:net:598:8)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)

Since Node Red is based on Node (v18 in my case), this should be consistent to say calling Node directly.  But your code never seems to reset or recover from the time out once it occurs?

**Comment by Jibun-no-Kage** [comment]
I think I might known why the timeout is happening... I am not exiting the async function correctly?  If I put an explicit 'return' in the async function (then) handler, does that release the allocated 'TV' object?

**Comment by Jibun-no-Kage** [comment]
Yeah, if invoked via node... have to call process.exit() at end of connect async function.  If calling from Node Red appears have to invoke 'return' to dispose of the TV object, but still testing.  Not a TypeScript or JavaScript expert, so I maybe completely wrong on how to dispose of the object correctly when invoked via Node Red.

**Comment by Jibun-no-Kage** [comment]
Well, appears I was mistaken, the timeout when TV object created/invoked within Node Red, still results ultimately in a timeout as documented above, regardless of how I try release the object.

**Comment by WesSouza** [comment]
I enabled discussions and moved your issues here, hope that's ok.

I'm not sure how Node Red actually works. Can you provide an example of how to set it up and how you are using it?

**Comment by Jibun-no-Kage** [comment]
Node Red communicates with message objects passed from node to node...  so in a what is called a 'function' node I can write raw JavaScript.  So I get a message into my function node, and I can add/remove/change 'msg' properties and then send the message, i.e. msg, to the next node in the chain/path... say a node that updates a GUI element on a dashboard.

```
//
const BOOLEAN = 'boolean',
      POWER = 'Power',
      SCREEN = 'Screen',
      MUTE = 'Mute',
      ERROR = 'Error',
      PERIOD = '.',
      QUESTION = '?',
      ON = 'On',
      OFF = 'Off',
      STATE = 'State';

const HUNDRED = 100,
      FIVE = 5,
      ONE = 1,
      ZERO  = 0,
      DELAY = 600;

//
const delay = ms => new Promise(res => setTimeout(res, ms));

//
const { LGTV }= global.get('lg'),
      TV = new LGTV(msg.host, msg.mac, msg.key);

//
msg.topic = (msg.payload != QUESTION) ? msg.topic : STATE;

//
TV.connect()

        .then(async () => {

                //
                switch (msg.topic) {

                        case SCREEN:

                                //
                                switch (msg.payload) {

                                        case false:

                                                //
                                                //await TV.setScreenMute('videomuteon')
                                                await TV.setScreenMute('screenmuteon')

                                                break;

                                        case true:

                                                //
                                                await TV.setScreenMute('allmuteoff')

                                                break;

                                }

                                flow.set(SCREEN, msg.payload)

                                break;


                        default:

                                //
                                node.warn(msg)

                                break;

                }

                //
                node.send(msg)
                
                //
                await TV.disconnect()
                
        })
.catch(console.error);

```

Once I added the explicit disconnect, I have yet to see the socket timeout issue, which makes sense after I reviewed the source code, and realized I was not calling an explicit disconnect.

The only issue I have now, is the following, which seem to be a soft error at this point...

Error: connect ECONNREFUSED 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761
}

I am calling getCurrentApp() about 10 seconds after the last direct command is issued, to update the status on the GUI i.e. the Node Red dashboard.  So if I enable mute for example, 10 seconds later, getCurrentApp() called.  My TV is configured both with wired and wireless connectivity, the 192.168.1.87 ip is the wired connection, BTW.

I know this error (-111) can mask other issues, such as the LG TV dropping network communication after power down.  Which then this error would make sense, given the LG TVs drop the network stack... requires WOL packet to power back on.   I just need to setup a testing scenario to see if this is what is happening.

**Comment by Jibun-no-Kage** [comment]
I explicitly powered off the TV via lgtv-ip-control, while watching a ping loop... and about 2 minutes after power down, the IP stack dropped, i.e. 192.168.1.87 pings stopped responding, typical behavior.  A bit later my getCurrentApp delayed query fired and I saw the following in the Node Red Log...


TimeoutError
    at Socket.handleTimeout (/root/node_modules/lgtv-ip-control/dist/classes/TinySocket.cjs:38:16)
    at Object.onceWrapper (node:events:631:28)
    at Socket.emit (node:events:517:28)
    at Socket._onTimeout (node:net:598:8)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
Error: connect EHOSTUNREACH 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -113,
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761
}
Error: connect EHOSTUNREACH 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -113,
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761
}
Error: connect EHOSTUNREACH 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -113,
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761
}
Error: connect EHOSTUNREACH 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -113,
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761
}
Error: connect EHOSTUNREACH 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -113,
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761
}
Error: connect EHOSTUNREACH 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -113,
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761

So the -111 error result is a bit of mystery.  Error -113 makes sense given the test method.  Not sure why I get -111, the communication is working and responsive when TV is poweron.  I send command to mute, it mutes, would the -111 result not block the mute command, i.e. it fails to complete?  Or does your code do multiple retires or something?

**Comment by Jibun-no-Kage** [comment]
When I powered on the TV... via WOL packet... then I started seeing the following in the Node Red Log...

Error: connect ECONNREFUSED 192.168.1.87:9761
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16) {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '192.168.1.87',
  port: 9761
}

I waited a few seconds, then the TV started responding to commands as expected.  So there seems to be a delay between power on and when the TV will accept commands via IP control?  Say the encryption/authentication program logic is not fully up, but the IP stack is, after power on?

I guess what I need to do, is implement in Node Red programming logic either way to test the TV IP stack is fully up, that the encryption/authentication logic is awake? So to speak?  I could disable the GUI controls until the TV is fully online after power up.  Really don't believe this is an issue with your code... at this point, unless you know or see an scenario that might indicate such? :)

**Comment by WesSouza** [comment]
I was finally able to install Node RED on a virtual machine to give it a spin and noticed that it can crash if a function produces an unhandled error.

Any connection issues or failures to read data or send commands can generate a promise rejection, which in turn will emit an unhandled error in the runtime that will stop Node RED.

```
24 Dec 18:06:29 - [info] Started flows
24 Dec 18:06:37 - [red] Uncaught Exception:
24 Dec 18:06:37 - [error] Error
    at Socket.handleTimeout (/home/wes/.node-red/node_modules/lgtv-ip-control/dist/classes/TinySocket.cjs:38:16)
```

From what I understood from your issues, we might want to add some sort of convenience method to wait until the TV starts responding, retrying until some point, with a possible timeout #112.

I've also identified a situation where the TV being off will be connectable, but not respond correctly to commands and result in `ResponseParseError`s #113.

I'll also work a bit more on the reliability of TinySocket, as it seems there are use cases we are not properly handling #114.

Until those are in place, you need to manually retry the `connect` method, making sure to catch:

```js
// Synchronous JS with Promises
tv.connect().catch(retryFunction);

// Asynchronous JS
try {
  await tv.connect();
} catch (error) {
  retryFunction(error);
}
```

**Comment by WesSouza** [comment]
@Jibun-no-Kage just released 4.1.0 with the `powerOnAndConnect` function which waits until the TV responds:

```js
const tv = new LGTV(/* options */);
await tv.powerOnAndConnect();
/* do stuff */
await tv.disconnect();
```

You can also pass a `maxRetries` to `connect` and it will quickly retry (which is what `powerOnAndConnect` uses):
```js
const tv = new LGTV(/* options */);
tv.powerOn();
await tv.connect({ maxRetries: 10 });
/* do stuff */
await tv.disconnect();
```

Let me know how that works for you.

**Comment by Jibun-no-Kage** [comment]
Very nice!  Thank you!  Looking forward to using this.

**Comment by Jibun-no-Kage** [comment]
Well, something odd happening.  I make sure to install 4.1.0 version in my node-red setup, but get an error that it cannot be loaded by node-red and node-red crashes.

It is in the packages.json...

```
root@template:~/.node-red# cat package.json
{
  "name": "node-red-project",
  "description": "initially created for you by Node-RED 3.1.3",
  "version": "0.0.1",
  "private": true,
  "dependencies": {
    "i2c-bus": "^5.2.3",
    "lgtv-ip-control": "^4.1.0",
    "node-red-contrib-buffer-parser": "^3.2.2",
    "node-red-contrib-play-audio": "^2.5.0",
    "node-red-node-pi-gpio": "^2.0.6",
    "node-red-node-ping": "^0.3.3",
    "node-red-node-random": "^0.4.1",
    "node-red-node-serialport": "^2.0.2",
    "node-red-node-smooth": "^0.1.2",
    "os": "^0.1.2"
  }
}
```
If is configured to preload in the node-red settings.js file...

```
    functionGlobalContext: {
        os:require('os'),
        i2c:require('i2c-bus'),
        lgtv:require('lgtv-ip-control')
    },
```

If I start node red...

root@template:~/.node-red# node-red-start

Start Node-RED

Once Node-RED has started, point a browser at http://192.168.1.7:1880
On Pi Node-RED works better with the Firefox or Chrome browser

Use   sudo systemctl enable nodered.service  to autostart Node-RED at every boot
Use   sudo systemctl disable nodered.service to disable autostart on boot

To find more nodes and example flows - go to http://flows.nodered.org

Starting as root systemd service.
Error loading settings file: /root/.node-red/settings.js
nodered.service: Main process exited, code=exited, status=1/FAILURE
nodered.service: Failed with result 'exit-code'.
nodered.service: Consumed 3.005s CPU time.


If I remove the reference in settings.js, nod-red starts...

```
    functionGlobalContext: {
        os:require('os'),
        i2c:require('i2c-bus'),
//      lgtv:require('lgtv-ip-control')
    },

Starting as root systemd service.
30 Dec 18:56:37 - [info]
Welcome to Node-RED
===================
30 Dec 18:56:37 - [info] Node-RED version: v3.1.3
30 Dec 18:56:37 - [info] Node.js  version: v18.19.0
30 Dec 18:56:37 - [info] Linux 6.1.21-v8+ arm64 LE
30 Dec 18:56:39 - [info] Loading palette nodes
30 Dec 18:56:43 - [info] Settings file  : /root/.node-red/settings.js
30 Dec 18:56:43 - [info] Context store  : 'memoryOnly' [module=memory]
30 Dec 18:56:43 - [info] Context store  : 'persistent' [module=localfilesystem]
30 Dec 18:56:43 - [info] User directory : /root/.node-red
30 Dec 18:56:43 - [warn] Projects disabled : editorTheme.projects.enabled=false
30 Dec 18:56:43 - [info] Flows file     : /root/.node-red/flows.json
30 Dec 18:56:43 - [info] Creating new flow file
30 Dec 18:56:43 - [warn] Encrypted credentials not found
30 Dec 18:56:43 - [info] Server now running at http://127.0.0.1:1880/
30 Dec 18:56:43 - [info] Starting flows
30 Dec 18:56:43 - [info] Started flows
```

So I tried to run the simple read me file example...

```
root@template:~/.node-red# node lgtv-mute-test.mjs
node:internal/errors:496
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'tslib' imported from /root/.node-red/node_modules/lgtv-ip-control/dist/classes/TinySocket.js
    at new NodeError (node:internal/errors:405:5)
    at packageResolve (node:internal/modules/esm/resolve:916:9)
    at moduleResolve (node:internal/modules/esm/resolve:973:20)
    at defaultResolve (node:internal/modules/esm/resolve:1193:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:403:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:372:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:249:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:39)
    at link (node:internal/modules/esm/module_job:75:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v18.19.0
```

Which is odd, because the module is installed...

```
node-red-project@0.0.1 /root/.node-red
├── i2c-bus@5.2.3
├── lgtv-ip-control@4.1.0
├── node-red-contrib-buffer-parser@3.2.2
├── node-red-contrib-play-audio@2.5.0
├── node-red-node-pi-gpio@2.0.6
├── node-red-node-ping@0.3.3
├── node-red-node-random@0.4.1
├── node-red-node-serialport@2.0.2
├── node-red-node-smooth@0.1.2
└── os@0.1.2
```
So the npm install of lgtv-ip-control does not pull in tslib module?

```
root@template:~/.node-red# npm install tslib@latest

added 1 package, and audited 36 packages in 3s

14 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

Loaded the missing tslib module, mjs test file runs.  If this is a dependency, should it not be pulled in on install?

**Reply by WesSouza** [reply]
Good catch, `tslib` became a dependency once I used native JS private properties, just released [v4.1.1](https://github.com/WesSouza/lgtv-ip-control/releases/tag/v4.1.1) with a fix.

**Comment by Jibun-no-Kage** [comment]
Oh...

```
root@template:~/.node-red# npm list
node-red-project@0.0.1 /root/.node-red
├── i2c-bus@5.2.3
├── lgtv-ip-control@4.1.0
├── node-red-contrib-buffer-parser@3.2.2
├── node-red-contrib-play-audio@2.5.0
├── node-red-node-pi-gpio@2.0.6
├── node-red-node-ping@0.3.3
├── node-red-node-random@0.4.1
├── node-red-node-serialport@2.0.2
├── node-red-node-smooth@0.1.2
├── os@0.1.2
└── tslib@2.6.2
```

Add... in your readme example... maybe include disconnect() and process.exit() so example actually completes execution?

```
root@template:~/.node-red# cat lgtv-mute-test.mjs
import { LGTV } from 'lgtv-ip-control';

const tv = new LGTV('192.168.1.87', 'XX:XX:XX:XX:XX:XX', 'XXXXXXXX');
// 192.168.1.88, XX:XX:XX:XX:XX:XX

tv.connect()
  .then(async () => {
    console.log('Mute...');
    await tv.setVolumeMute(true);
    console.log('Mute...');
    await tv.setVolumeMute(false);
    tv.disconnect()
    process.exit()
  })
  .catch(console.error);
```
Or is there a more graceful way to exit from the 'then' promise handler?

**Reply by WesSouza** [reply]
As far as I know `tv.disconnect` should be enough.

Given you are inside of a promise, and tv.disconnect returns a promise, you can either:

```js
// Await the disconnect
tv.connect()
  .then(async () => {
    // ...
    await tv.disconnect()
  })

// Return the disconnect (it will chain to the `catch` if it fails)
tv.connect()
  .then(async () => {
    // ...
    return tv.disconnect()
  })

// Use a `finally` after the catch
tv.connect()
  .then(async () => {
    // ...
  })
  .catch(console.error)
  .finally(() => tv.disconnect());
```

The `finally` option is the best because in theory you still should try to disconnect in case of an error, and would need to add the code on both `then` and `catch` for that.

In terms of `process.exit`, it should not be needed as long as you disconnect, otherwise you are force quitting the process and leaving the socket open.

I've updated the README.md example to add the disconnect as it was indeed missing.

**Reply by WesSouza** [reply]
If you want to emit an exit code in case of an error using promises and `then..catch..finally`, you can pass the result to the finally block as such:


```js
tv
  .connect()
  .then(async () => {
    // ...
    return true;
  })
  // Log any errors
  .catch((error) => {
    console.error(error);
    return false;
  })
  // Tries disconnecting once done
  .finally((hadError) => {
    tv.disconnect();
    if (hadError) {
      process.exit(1);
    }
  });
```

If there is no error in the `then` block, it will go to `finally`. The first argument will be whatever was returned by the previous function.

**Comment by Jibun-no-Kage** [comment]
Closing since was resolved.
